File: Cpp/Motherboard.cpp
#pragma once
#include <iostream>
#include <string>
#include "Komponen.cpp"
using namespace std;

// Interface untuk USB ports (digunakan untuk multiple inheritance)
class USBInterface {
protected:
    int usb2Ports;
    int usb3Ports;

public:
    USBInterface() {
        this->usb2Ports = 0;
        this->usb3Ports = 0;
    }

    USBInterface(int usb2Ports, int usb3Ports) {
        this->usb2Ports = usb2Ports;
        this->usb3Ports = usb3Ports;
    }

    void setUsb2Ports(int usb2Ports) {
        this->usb2Ports = usb2Ports;
    }

    void setUsb3Ports(int usb3Ports) {
        this->usb3Ports = usb3Ports;
    }

    int getUsb2Ports() const {
        return this->usb2Ports;
    }

    int getUsb3Ports() const {
        return this->usb3Ports;
    }

    virtual void displayUSBInfo() const {
        cout << "USB 2.0 Ports: " << this->usb2Ports << endl;
        cout << "USB 3.0 Ports: " << this->usb3Ports << endl;
    }

    virtual ~USBInterface() {}
};

// Motherboard menggunakan multiple inheritance
class Motherboard : public Komponen, public USBInterface {
private:
    string chipset;
    string formFactor;  // ATX, MicroATX, etc
    int ramSlots;
    int m2Slots;

public:
    Motherboard() : Komponen(), USBInterface() {
        this->chipset = "";
        this->formFactor = "";
        this->ramSlots = 0;
        this->m2Slots = 0;
    }

    Motherboard(string chipset, string formFactor, int ramSlots, int m2Slots, 
                int usb2Ports, int usb3Ports, string merk, string nama, int harga) 
        : Komponen(merk, nama, harga), USBInterface(usb2Ports, usb3Ports) {
        this->chipset = chipset;
        this->formFactor = formFactor;
        this->ramSlots = ramSlots;
        this->m2Slots = m2Slots;
    }

    void setChipset(string chipset) {
        this->chipset = chipset;
    }

    void setFormFactor(string formFactor) {
        this->formFactor = formFactor;
    }

    void setRamSlots(int ramSlots) {
        this->ramSlots = ramSlots;
    }

    void setM2Slots(int m2Slots) {
        this->m2Slots = m2Slots;
    }

    string getChipset() const {
        return this->chipset;
    }

    string getFormFactor() const {
        return this->formFactor;
    }

    int getRamSlots() const {
        return this->ramSlots;
    }

    int getM2Slots() const {
        return this->m2Slots;
    }

    void displayInfo() const override {
        Komponen::displayInfo();
        cout << "Chipset: " << this->chipset << endl;
        cout << "Form Factor: " << this->formFactor << endl;
        cout << "RAM Slots: " << this->ramSlots << endl;
        cout << "M.2 Slots: " << this->m2Slots << endl;
        displayUSBInfo();
    }

    void displayUSBInfo() const override {
        USBInterface::displayUSBInfo();
    }

    ~Motherboard() {}
};

File: Python/Motherboard.py
from Komponen import Komponen

# Interface untuk USB ports (digunakan untuk multiple inheritance)
class USBInterface:
    def __init__(self, usb2_ports=0, usb3_ports=0):
        self.usb2_ports = usb2_ports
        self.usb3_ports = usb3_ports
        
    def set_usb2_ports(self, usb2_ports):
        self.usb2_ports = usb2_ports
        
    def set_usb3_ports(self, usb3_ports):
        self.usb3_ports = usb3_ports
        
    def get_usb2_ports(self):
        return self.usb2_ports
        
    def get_usb3_ports(self):
        return self.usb3_ports
        
    def display_usb_info(self):
        print(f"USB 2.0 Ports: {self.usb2_ports}")
        print(f"USB 3.0 Ports: {self.usb3_ports}")
        
# Motherboard menggunakan multiple inheritance
class Motherboard(Komponen, USBInterface):
    def __init__(self, chipset="", form_factor="", ram_slots=0, m2_slots=0, 
                 usb2_ports=0, usb3_ports=0, merk="", nama="", harga=0):
        Komponen.__init__(self, merk, nama, harga)
        USBInterface.__init__(self, usb2_ports, usb3_ports)
        self.chipset = chipset
        self.form_factor = form_factor
        self.ram_slots = ram_slots
        self.m2_slots = m2_slots
        
    def set_chipset(self, chipset):
        self.chipset = chipset
        
    def set_form_factor(self, form_factor):
        self.form_factor = form_factor
        
    def set_ram_slots(self, ram_slots):
        self.ram_slots = ram_slots
        
    def set_m2_slots(self, m2_slots):
        self.m2_slots = m2_slots
        
    def get_chipset(self):
        return self.chipset
        
    def get_form_factor(self):
        return self.form_factor
        
    def get_ram_slots(self):
        return self.ram_slots
        
    def get_m2_slots(self):
        return self.m2_slots
        
    def display_info(self):
        Komponen.display_info(self)
        print(f"Chipset: {self.chipset}")
        print(f"Form Factor: {self.form_factor}")
        print(f"RAM Slots: {self.ram_slots}")
        print(f"M.2 Slots: {self.m2_slots}")
        self.display_usb_info()
